[Merged by Bors] -  feat(group_theory/torsion): define torsion groups

Julian opened the pull request:
I grepped for torsion group and didn't find anything -- hopefully adding this makes sense here.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

alexjbest commented on src/group_theory/exponent.lean:
```diff
@@ -52,19 +53,40 @@ variables (G) [monoid G]
 @[to_additive "A predicate on an additive monoid saying that there is a positive integer `n` such
   that `n • g = 0` for all `g`."]
-def exponent_exists  := ∃ n, 0 < n ∧ ∀ g : G, g ^ n = 1
+def is_torsion := ∃ n, 0 < n ∧ ∀ g : G, g ^ n = 1
```
A torsion group should be one where every element is torsion, which isn't the same as the exponent existing. For instance the direct sum of cyclic groups of all possible finite orders is torsion but the exponent doesnt exist.

Julian replied:
Thanks for correcting my mistake Alex, and for the example. I'm working on defining these separately and leaving what's here as-is then.

Julian committed:
Define torsion groups.

Subgroups of torsion groups are torsion, and quotients of torsion
groups are torsion.

Julian committed:
exponent_exists -> is_torsion

Julian committed:
Finite groups are torsion groups.

Julian committed:
Redefine torsion in terms of is_of_fin_order.

Julian-Kuelshammer commented on src/group_theory/torsion.lean:
```diff
@@ -0,0 +1,75 @@
+  intro g,
+  obtain ⟨n, ⟨npos, hn⟩⟩ := tG g,
+  refine ⟨n, npos, subtype.coe_injective _⟩,
+  rw is_periodic_pt_mul_iff_pow_eq_one ↑g at hn,
```
For me it would be more natural to work entirely in the group API. Would it be much harder to use `is_of_fin_order_iff_pow_eq_one` instead?

Julian replied:
Oh! This definitely looks easier, somehow I missed it. Thank you! Will give those changes a shot.

Julian-Kuelshammer commented on src/group_theory/torsion.lean:
```diff
@@ -0,0 +1,75 @@
+λ g, quotient.induction_on' g $ λ a, begin
+  obtain ⟨n, ⟨npos, hn⟩⟩ := tG a,
+  refine ⟨n, npos, (is_periodic_pt_mul_iff_pow_eq_one _).mpr $ (quotient_group.con N).eq.mpr _⟩,
+  exact (is_periodic_pt_mul_iff_pow_eq_one a).mp hn ▸ (quotient_group.con N).eq.mp rfl,
```
A similar comment applies here.

Julian-Kuelshammer commented on src/group_theory/torsion.lean:
```diff
@@ -0,0 +1,75 @@
+lemma exponent_exists.is_torsion (h : exponent_exists G) : is_torsion G := begin
+  intro g,
+  obtain ⟨n, ⟨npos, hn⟩⟩ := h,
+  exact ⟨n, npos, (is_periodic_pt_mul_iff_pow_eq_one _).mpr $ hn g⟩,
```
And here.

Julian-Kuelshammer commented on src/group_theory/torsion.lean:
```diff
@@ -0,0 +1,75 @@
+* `add_monoid.is_torsion` the additive version of `monoid.is_torsion`.
+
+## Main results
+
```
A TODO for later could be to define the torsion subgroup of a group (see e.g. [here](https://encyclopediaofmath.org/wiki/Torsion_group)), prove that it is a normal subgroup and define the torsion-free quotient group.

Julian replied:
Thanks! Makes sense, can do -- so far I found/was reading "A Course in the Theory of Groups" (Robinson) which also has a few sections mentioning these (and looked like also has some other kinds of groups defined in terms of them).

Julian replied:
(I added your suggestions as notes -- hopefully that was the right thing to do.)

Also I see there are some things mentioning torsion-free modules defined in `linear_algebra/free_module/pid.lean` -- are these related / if so presumably I should try to generalize what's there from a ring to the additive group?

Julian-Kuelshammer replied:
An `add_comm_group` is the same as a Z-module. In this case the concept of a torsion (sub)group and of a torsion (sub)module agree. So the two concepts are generalisations in two different directions (either generalising by removing the `comm` condition or by generalising the ring.

Julian replied:
Aha, I see, thanks!

Julian committed:
Apply the better iff suggestion from code review.

Julian committed:
One more.

Julian committed:
Add the TODOs

Julian committed:
Add the other direction of is_torsion -> exponent_exists

Julian committed:
Rest of the missing to_additives

Vierkantor commented on src/group_theory/torsion.lean:
```diff
@@ -0,0 +1,86 @@
+
+/--Subgroups of torsion groups are torsion groups. -/
+@[to_additive "Subgroups of additive torsion groups are additive torsion groups."]
+lemma subgroup.is_torsion {tG : is_torsion G} (H : subgroup G) : is_torsion H := begin
```
Let's make `tG` explicit and enable dot notation:
```suggestion
lemma is_torsion.subgroup (tG : is_torsion G) (H : subgroup G) : is_torsion H := begin
```

Vierkantor commented on src/group_theory/torsion.lean:
```diff
@@ -0,0 +1,86 @@
+
+/--Quotient groups of torsion groups are torsion groups. -/
+@[to_additive "Quotient groups of additive torsion groups are additive torsion groups."]
+lemma quotient_group.is_torsion [nN : N.normal] (tG : is_torsion G) : is_torsion (G ⧸ N) :=
```
```suggestion
lemma is_torsion.quotient_group [nN : N.normal] (tG : is_torsion G) : is_torsion (G ⧸ N) :=
```

Vierkantor commented on src/group_theory/torsion.lean:
```diff
@@ -0,0 +1,86 @@
+
+/--The group exponent exists for any bounded torsion group. -/
+@[to_additive exponent_exists.of_is_add_torsion]
+lemma exponent_exists.of_is_torsion
```
```suggestion
lemma is_torsion.exponent_exists
```

Vierkantor commented on src/group_theory/torsion.lean:
```diff
@@ -0,0 +1,86 @@
+@[to_additive exponent_exists.of_is_add_torsion]
+lemma exponent_exists.of_is_torsion
+  (tG : is_torsion G)
+  (bounded : (set.range (λ g : G, order_of g)).finite) : exponent_exists G :=
```
I prefer having the conclusion on a separate line so it stands out more:
```suggestion
  (tG : is_torsion G)   (bounded : (set.range (λ g : G, order_of g)).finite) :
  exponent_exists G :=
```

Julian committed:
Apply suggestions from code review

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

Vierkantor reviewed (approved):
Thanks!

bors d=@julian-kuelshammer

Julian committed:
Fix the to-additive name of one, and rename another.

Julian-Kuelshammer commented on src/group_theory/torsion.lean:
```diff
@@ -0,0 +1,86 @@
+λ g, quotient.induction_on' g $ λ a, begin
+  rw is_of_fin_order_iff_pow_eq_one,
+  obtain ⟨n, ⟨npos, hn⟩⟩ := (is_of_fin_order_iff_pow_eq_one _).mp (tG a),
+  exact ⟨n, npos, (quotient_group.con N).eq.mpr (hn ▸ (quotient_group.con N).eq.mp rfl)⟩,
```
Now that I'm looking at this again. Would it make sense to deduce that from some inheritance property of `is_of_fin_order`, that says something like: An element of a subgroup is of finite order if and only if it is of finite order in the big group? And similarly, if an element is of finite order in a big group, then it is of finite order in any quotient? This should basically have the same proof, just without the `intro g` part.

Julian replied:
I'll give those a shot, thanks!

Julian replied:
OK I believe I did what was suggested (and indeed things turn out nicer I think). Let me know how this looks. Thanks again.

Julian-Kuelshammer replied:
LGTM.

Julian committed:
Revert the incorrect additional rename.

Julian committed:
Prove things via elementwise deductions from is_of_fin_order.

Julian commented on src/group_theory/order_of_element.lean:
```diff
@@ -68,6 +69,23 @@ lemma is_of_fin_order_iff_pow_eq_one (x : G) :
 
+/-- Elements of finite order are of finite order in subgroups.-/
+@[to_additive is_of_fin_add_order_iff_coe]
+lemma is_of_fin_order_iff_coe {G : Type u} [group G] (H : subgroup G) (x : H) :
```
(I am unsure whether I've named these 2 properly, so comments welcome there of course.)
